### ISSUE TITLE
Consider switching type checker

### ISSUE THREAD

**Issue by mccalluc** [opendp/opendp#1092]
Starting to look at
- #1067

and before sinking a lot of time into that (either refactoring code so mypy understands it, or googling esoteric error messages) I am wondering if another type checker would give us fewer false positives, while still capturing the real problems. For comparison, I just ran either on `mod.py`.

## `mypy`: Reference implementation
- Finds these errors:
```
src/opendp/mod.py:11: error: Unsupported dynamic base class "ctypes.POINTER"  [misc]
src/opendp/mod.py:180: error: Unsupported dynamic base class "ctypes.POINTER"  [misc]
src/opendp/mod.py:273: error: Name "PartialChain" is not defined  [name-defined]
src/opendp/mod.py:283: error: Too many arguments for "__call__"  [call-arg]
src/opendp/mod.py:376: error: Cannot assign to a type  [misc]
src/opendp/mod.py:404: error: Unsupported dynamic base class "ctypes.POINTER"  [misc]
src/opendp/mod.py:424: error: Unsupported dynamic base class "ctypes.POINTER"  [misc]
src/opendp/mod.py:467: error: Unsupported dynamic base class "ctypes.POINTER"  [misc]
src/opendp/mod.py:501: error: Unsupported dynamic base class "ctypes.POINTER"  [misc]
src/opendp/mod.py:750: error: Overloaded function signatures 1 and 2 overlap with incompatible return types  [overload-overlap]
Found 10 errors in 1 file (checked 1 source file)
```
(googling "Unsupported dynamic base class" is what prompted me to look at other options.)

## `pyright`: Microsoft
- Install is quick, with minimal dependencies, but when it runs there is log noise from npm, which is disconcerting.
- Docs explain that their use case is IDE support, and the focus is speed.
- Finds these errors:
```
/Users/chuck/github/opendp/opendp/python/src/opendp/mod.py
  /Users/chuck/github/opendp/opendp/python/src/opendp/mod.py:273:160 - error: "PartialChain" is not defined (reportUndefinedVariable)
  /Users/chuck/github/opendp/opendp/python/src/opendp/mod.py:940:16 - error: Expression of type "tuple[Unknown, ...] | None" cannot be assigned to return type "Tuple[float, float] | Tuple[int, int] | None" (reportGeneralTypeIssues)
  /Users/chuck/github/opendp/opendp/python/src/opendp/mod.py:963:12 - error: Expression of type "tuple[Unknown, ...] | None" cannot be assigned to return type "Tuple[float, float] | Tuple[int, int] | None" (reportGeneralTypeIssues)
3 errors, 0 warnings, 0 informations
```
(with nice highlighting!)

## `pytype`: Google
- A large number of transitive dependencies: ninja, typing-inspect, toml, tabulate, pyyaml, pyparsing, pycnite, networkx, MarkupSafe, attrs, pydot, libcst, jinja2, importlab, pytype
- Docs explain that the focus is on strong inference, even where typing isn't explict.
- Run is _very_ slow, particularly the first time, and it finds only this error:
```
File "/Users/chuck/github/opendp/opendp/python/src/opendp/mod.py", line 273, in Transformation: Name 'PartialChain' is not defined [name-error]
```

## `pyre`: Facebook
- Seems like it needs an extra initialization step, and even then it doesn't run for me out of the box: 
```
env: python.pyre: No such file or directory
```
Not auspicious.

---------------

I'll investigate pyright a bit more and either make a PR, or close this issue if staying with mypy makes more sense.

**Comment by Shoeboxam** [comment]
For context, I resolved errors/warnings raised by pyright in #738.

**Comment by mccalluc** [comment]
Michael and I talked just now and decided to stay with mypy.
- Either is fast enough for our purposes.
- Both are giving us warnings right now: Some warnings from both, and some are unique to a given tool -- So we'll probably get some false positives, and false negatives, with either.
- In some sense pyright may be better... but mypy still has a large market share, and the idioms it favors will not be surprising to the wider community.
- pyright's connection MS has been a concern for some: Not really OS, changes in specs that disadvantage other platforms, analytics reported back, etc.
